### PR TITLE
chore: operator tests

### PR DESCRIPTION
**Pull request by charleenfei**
Please review PR #173 first

### CODE REVIEW

File: src/core/lender/tranche/operator.sol [hunk 1]
```diff
@@ -21,6 +21,7 @@ import "ds-math/math.sol";
 contract TrancheLike {
     function supply(address usr, uint currencyAmount, uint tokenAmount) public;
     function redeem(address usr, uint currencyAmount, uint tokenAmount ) public;
+    function mock(uint t) public;
```
Comment (lucasvo) [inline src/core/lender/tranche/operator.sol]: this should not be in there

File: src/core/lender/tranche/test/operator.t.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    function testWhitelistSupply() public {
+        assessor.setReturn("tokenPrice", 1 ether);
+        whitelist.relyInvestor(address(this));
+        whitelist.supply(100 ether);
+        assertEq(tranche.calls("supply"), 1);
+        assertEq(assessor.calls("tokenPrice"), 1);
+    }
+
+    function testWhitelistRedeem() public {
+        assessor.setReturn("tokenPrice", 1 ether);
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]:
```suggestion
        assessor.setReturn("tokenPrice", ONE);
```

File: src/core/lender/tranche/test/operator.t.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    function testFailAllowanceSupply() public {
+        assessor.setReturn("tokenPrice", 1 ether);
+        allowance.approve(address(this), 100 ether, 50 ether);
+        allowance.supply(100 ether);
+    }
+
+    function testFailAllowanceRedeem() public {
+        assessor.setReturn("tokenPrice", 1 ether);
+        allowance.approve(address(this), 50 ether, 100 ether);
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]: can we change the allowanceWhitelist to switch the order of parameters in the contract?

I implicitly applied the rule, if we have a currencyAmount and tokenAmount in a method. Always the currencyAmount first :)

my fault

File: src/core/lender/tranche/test/operator.t.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        assessor.setReturn("tokenPrice", 1 ether);
+        whitelist.relyInvestor(address(this));
+        whitelist.supply(100 ether);
+        assertEq(tranche.calls("supply"), 1);
+        assertEq(assessor.calls("tokenPrice"), 1);
+    }
+
+    function testWhitelistRedeem() public {
+        assessor.setReturn("tokenPrice", 1 ether);
+        whitelist.relyInvestor(address(this));
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]: We need to be aware we are testing here an edge case. 

`Investor == Admin`

maybe also a test with another address. We would need a investorContract with a `doSupply` and `doReedem`

File: src/core/lender/test/mock/assessor.sol [hunk 5]
```diff
@@ -1,14 +1,57 @@
...
+    function calcSeniorAssetValue(uint poolValue, uint trancheReserve, uint trancheDebt, uint juniorReserve) internal returns (uint) {
+        return ((poolValue + juniorReserve) >= trancheDebt) ? (trancheDebt + trancheReserve) : (poolValue + juniorReserve + trancheReserve);
+    }
+    function juniorReserve() internal returns (uint) {
+        return call("juniorReserve");
+    }
+    function seniorDebt() internal returns (uint) {
+        return call("seniorDebt");
     }
 }
```
Comment (manuelpolzhofer) [inline src/core/lender/test/mock/assessor.sol]: lets update it to mock behaviour

File: src/core/lender/tranche/test/operator.t.sol [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+pragma solidity >=0.4.23;
+
+import "ds-test/test.sol";
+
+import "../../test/mock/tranche.sol";
+import "../../test/mock/assessor.sol";
+import "../operator.sol";
+
+contract OperatorTest is DSTest {
```
[the remark this answers is not in the document]
Comment (charleenfei) [reply src/core/lender/tranche/test/operator.t.sol]: i feel like the depend method is so generic, if we want to test it we could maybe think about testing it in the tinlake-auth pkg. wdyt?

File: src/core/lender/tranche/test/operator.t.sol [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        whitelist.relyInvestor(address(this));
+        whitelist.redeem(100 ether);
+        assertEq(tranche.calls("redeem"), 1);
+        assertEq(assessor.calls("tokenPrice"), 1);
+    }
+
+    function testWhitelistSupplyInvestor() public {
+        whitelist.relyInvestor(address(investor));
+        investor.doSupply(address(whitelist), 100 ether);
+        assertEq(tranche.calls("supply"), 1);
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]: I think we should check here for the correct `currencyAmount` and `tokenAmount`

File: src/core/lender/tranche/test/operator.t.sol [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        assertEq(assessor.calls("tokenPrice"), 1);
+    }
+
+    function testWhitelistRedeemInvestor() public {
+        whitelist.relyInvestor(address(investor));
+        investor.doRedeem(address(whitelist), 100 ether);
+        assertEq(tranche.calls("redeem"), 1);
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]: same here

File: src/core/lender/tranche/test/operator.t.sol [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+    function testAllowanceSupply() public {
+        allowance.approve(address(this), 100 ether, 100 ether);
+        allowance.supply(100 ether);
+        assertEq(tranche.calls("supply"), 1);
+        assertEq(assessor.calls("tokenPrice"), 1);
+    }
+
+    function testAllowanceRedeem() public {
+        allowance.approve(address(this), 100 ether, 100 ether);
```
Comment (manuelpolzhofer) [inline src/core/lender/tranche/test/operator.t.sol]: lets use the investor contract here.
```suggestion
        allowance.approve(address(investor), 100 ether, 100 ether);
```
